### ISSUE TITLE
TAB auto completion with selectable completion

### ISSUE THREAD

**Issue by bw1faeh0** [chrisant996/clink#52]
[Feature request]
I enjoy using your clink variant!

But... If I use the TAB key when navigation through my directories or files I only get a list of matching items, but can't select them by tabbing through the list:
![image](https://user-images.githubusercontent.com/29235332/104280486-4db8d900-54ac-11eb-8c14-2be7e8e0cf97.png)

It would be nice if additional TAB-strokes would select item by item from the list, so I can tab two or three times to select my item and only hit enter, like in zsh:

![image](https://user-images.githubusercontent.com/29235332/104281203-6a094580-54ad-11eb-96a3-c4e12286a843.png)
(image taken after typing `cd nginx` and tabbing two times)
![image](https://user-images.githubusercontent.com/29235332/104281332-a50b7900-54ad-11eb-9776-0a9105e7ad70.png)
(image taken after typing `cd nginx` and tabbing four times)

**Comment by chrisant996** [comment]
Yes, that exists, and the Readline documentation covers it.

Bind the tab key to `menu-complete` or `old-menu-complete`. The section about Key Binding in clink.html should be helpful. If not, please share feedback about why not, so I can improve the documentation.

Personally, I prefer tab to cycle, for similarity with CMD. But Clink has always behaved like Readline and Bash, which don't cycle. I don't want to change the default because it would be disruptive to most existing users of Clink.

**Comment by bw1faeh0** [comment]
Binding menu-complete was very helpful, thanks for the hint!

One suggestion for your documentation: 
You should mention the location and name of the file where to put the key bindings in. I had to google to find out I have to use the .inputrc.

**Comment by chrisant996** [comment]
> One suggestion for your documentation:
> You should mention the location and name of the file where to put the key bindings in. I had to google to find out I have to use the .inputrc.

Thanks, I'll look into that.  What section of the Clink documentation did you refer to?  The "Key Bindings" section explicitly mentions `.inputrc` a few times, has a sample file (listing my personal key bindings), and has a link to the Readline documentation.

It doesn't mention how to find _where_ the .inputrc file is located, though...!  I'll update it to mention running `clink info` to find where the .inputrc file is searched for (there are several locations it looks).

**Comment by bw1faeh0** [comment]
I used your sample file and the Readline documentation but couldn't find where to put the file. From my point of view this is the missing part in you docu that I was googleing ;)

**Comment by chrisant996** [comment]
Perfect, thanks.

I think you mean that you weren't straight for configuring key bindings, without reading the section on "Configuring Readline", right?

That's of course perfectly reasonable! I just want to make sure that's what happened, because I think the best documentation fix is to make the Key Bindings section more strongly and directly send the reader over to the Configuring Readline section. To me, that section starts with a pretty detailed explanation of the file locations, though I see a couple clarifications I want to add to it.

Would the Configuring Readline section have given you the needed info?

Thanks again!

**Comment by ganego** [comment]
Have been using clink for years and this was always really annoying that I cannot cycle TAB completion (at least one thing cmd.exe had right).  

What is the difference between `old-menu-complete` & `menu-complete`. They seem to do the same?  

Anyway, two suggestions:  
- `clink-popup-complete` should loop (meaning the first entry is selected, I press arrow up and it selects the last entry).  
- New way for TAB complete to work (and how I have it on bash): First TAB press shows a list of all available completion suggestions. All consecutive TAB presses cycle the commands. So basically the first TAB press should do what `complete` does that in your `.inputrc` example is set to `Ctrl+Space`. With the possibility to cycle suggestions backwards.

Thank you for updating clink!

**Comment by chrisant996** [comment]
> What is the difference between `old-menu-complete` & `menu-complete`. They seem to do the same?

`old-menu-complete` is closest to how CMD works, and it's what I have bound to my <kbd>Tab</kbd> key.

According to comments in the Readline source code, the differences are:
```cpp
/* The current version of menu completion.
   The differences between this function and the original are:

1. It honors the maximum number of completions variable (completion-query-items)
2. It appends to the word as usual if there is only one match
3. It displays the common prefix if there is one, and makes it the first menu
   choice if the menu-complete-display-prefix option is enabled
*/
 
int
rl_menu_complete (int count, int ignore)
```

> * `clink-popup-complete` should loop (meaning the first entry is selected, I press arrow up and it selects the last entry).

Yes, fair.  What needs to happen is add a "loop around" config variable in Readline that affects the whole `menu-complete` family of commands, including `clink-popup-complete`.  I actually hate the cycling around, and it's already on my ToDo list to add such a config variable, and in the meantime I made the popup version behave how I prefer.

> * New way for TAB complete to work (and how I have it on bash) - I mean this like an option like `menu-complete-2` or so: First TAB press shows a list of all available completion suggestions. All consecutive TAB presses cycle the commands.

Yes, the menu-select style of interactive completion in fish and zsh, etc.  It's one of the next things on my list.  It may or may not make it into the first official release from the chrisant996 fork.

FWIW:  I'm currently hovering in a triage mode on the way to an official release.  If any "Must" fix comes up, then I also roll in any "Like" fixes/features that are ready.  That's why there are still release candidates coming -- we're still collectively finding "Must" fixes frequently enough that it isn't ready to publish an official release yet.

**Comment by chrisant996** [comment]
There is a new `clink-select-complete` command.  It isn't released yet, but if you build Clink locally then the command is available.  There are still a few things to finish, but the command is basically functional.

**Comment by chrisant996** [comment]
Published in [v1.2.19](https://github.com/chrisant996/clink/releases/tag/v1.2.19).
